src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs

```diff
@@ -118,6 +122,224 @@ public static class PrereleaseCompatibilityTransformer
             return transformedSarifLog;
         }
 
+        private static bool ApplyChangesFromTC33(JObject sarifLog)
+        {
+            UpdateSarifLogVersion(sarifLog);
+            if (sarifLog["runs"] is JArray runs)
+            {
+                foreach (JObject run in runs)
+                {
+                    // https://github.com/oasis-tcs/sarif-spec/issues/340
+                    AddLogicalLocationToAllLocationNodes(run);
+                }
+            }
+            return true;
+        }
+
+        private static void AddLogicalLocationToAllLocationNodes(JObject run)
+        {
+            // We need to remove location.fullyQualifiedLogicalName (string) and location.logicalLocationIndex (int) and transfer those
+            // to a single LogicalLocation object with properties "fullyQualifiedName" and "parentIndex".
+
+            // Previously:
+            //      "location" : {
+            //          "fullyQualifiedLogicalName" : "test" ,
+            //          "logicalLocationIndex" : 1
+            //      }
+            // Now:
+            //      "location" : {
+            //          "logicalLocation" : {
+            //              "fullyQualifiedName" : "test",
+            //              "parentIndex" : 1
+            //          }
+            //      }
+
+            // The code walks through all possible paths to 'location' node and performs updates.
+
+            // sarifLog.runs[].results[].locations[]
```

Comment:
>sarifLog.runs[].results[].locations[] [](start = 15, length = 37)

i tried to write a common method for getting nodes by passing path. was running into few challenges, we ll discuss offline :) for now i have followed the old approach. this one was very unpleasant :(.

Reply:
ok, let's talk tomorrow.

---
In reply to: [264979458](https://github.com/Microsoft/sarif-sdk/pull/1347#discussion_r264979458) [](ancestors = 264979458)

Reply:
will do, this pr has been a bit big, so i was focusing today on other PRs. Lets chat

---
In reply to: [264982352](https://github.com/Microsoft/sarif-sdk/pull/1347#discussion_r264982352) [](ancestors = 264982352,264979458)